New Swift client (wrapping C++ client)

bhamiltoncx opened the pull request:
This is an alternative to https://github.com/google/myanmar-tools/pull/50 .

Instead of introducing a new Swift client with an Xcode project, this CL:

1. Exposes the existing C++ client with a C interface, and
2. Creates a Swift wrapper for the C interface using the Swift Package Manager (SwiftPM).

To build:

```
make client-swift
```

To test:

```
make client-swift-test
```

Some caveats:

1) Everything is linked dynamically currently. I tried making a static library of the C++ `libmyanmartools`, but I couldn't figure out how to convince CMake to link glog statically.
2) I didn't include all the tests from https://github.com/google/myanmar-tools/pull/50 . If this looks good, I'm happy to merge them in.

lawinko commented:
@bhamiltoncx, That is a great idea. I was thinking to wait for Swift 5.3 to get the resource import feature in Swift Package Manager. Because of not to depend on the iOS/macOS platform.

bhamiltoncx commented:
> @bhamiltoncx, That is a great idea. I was thinking to wait for Swift 5.3 to get the resource import feature in Swift Package Manager. Because of not to depend on the iOS/macOS platform.

I think this CL should not depend on iOS or macOS in any way. I believe it should work OK with Swift on Linux at least. (I'm not sure about Swift on Windows.)

lawinko commented:
@bhamiltoncx  yes, you are right. I did notice that SPM doesn't support resource import after all the codes are complete hehe. That's why I chose COCOAPODS to use as a solution. I think your way is the best right now.

sffc commented on clients/cpp/zawgyi_detector.cpp:
```diff
@@ -275,3 +275,13 @@ double ZawgyiDetector::GetZawgyiProbability(const char* input_utf8,
+}
+
+double GMTGetZawgyiProbabilityWithLength(const char* input_utf8, int32_t length) {
+  ZawgyiDetector detector;
```
Suggestion (can ignore if too hard): The constructor does a bit of work (parsing the model).  It would be nicer if you didn't have to construct the ZawgyiDetector each time.  The pattern ICU uses for this is an opaque struct that you open and close, like this:

```c
GMTZawgyiDetector detector = gmt_openZawgyiDetector();
double result = gmt_getZawgyiProbability(detector, "...");
gmt_closeZawgyiDetector(detector);
```

bhamiltoncx replied:
Done, thanks!

bhamiltoncx commented:
> Thank you!
> 
> Yes, I like this PR. It will reduce maintenance costs. I gave one API suggestion. Beyond that, it would be great to see C tests, as well as the data-driven Swift tests.

Thanks! I updated the API, added C tests, and data-driven Swift tests.

The tests uncovered a bug in Swift itself which I filed as https://bugs.swift.org/browse/SR-13047 (comparing `-Infinity` to itself fails when passing an `accuracy` value). I included a workaround for the bug.

bhamiltoncx commented:
Argh. I thought "close" would close the comment, not the PR. Reopening. :)

sffc commented on clients/cpp/public/myanmartools.h:
```diff
@@ -35,4 +40,24 @@ class ZawgyiDetector {
+#endif  // __cplusplus
+
+// C bindings (the "GMT" prefix stands for "Google Myanmar Tools")
+typedef void *GMTZawgyiDetector;
```
Suggestion: if you use

```c
typedef struct GMTZawgyiDetector GMTZawgyiDetector;
```

then your C code uses the type `GMTZawgyiDetector*`, and you get more type safety.  It also makes it more clear that this is a pointer that needs to be cleaned up.

https://stackoverflow.com/a/3965308/1407170

bhamiltoncx replied:
Done, thanks for the tip.
